cpp/src/arrow/testing/random_test.cc

```diff
@@ -0,0 +1,195 @@
+                             field("int8", int8()),
+                             field("int16", int16()),
+                             field("int32", int32()),
+                         })),
+    field("denseunion", dense_union({
+                            field("int8", int8()),
+                            field("int16", int16()),
+                            field("int32", int32()),
+                        })),
+    field("dictionary", dictionary(int8(), utf8())), field("map", map(int8(), utf8())),
+    field("fixedsizelist", fixed_size_list(int8(), 4)),
+    field("durationns", duration(TimeUnit::NANO)), field("largestring", large_utf8()),
+    field("largebinary", large_binary()),
+    field("largelistlistint8", large_list(list(int8()))));
+
+INSTANTIATE_TEST_SUITE_P(
+    TestRandomArrayGeneration, RandomArrayTest, values,
+    [](const ::testing::TestParamInfo<RandomArrayTest::ParamType>& info) {
+      return std::to_string(info.index) + info.param->name();
+    });
+
+using NumericTypes =
+    ::testing::Types<UInt8Type, Int8Type, UInt16Type, Int16Type, UInt32Type, Int32Type,
+                     HalfFloatType, FloatType, DoubleType>;
+TYPED_TEST_SUITE(RandomNumericArrayTest, NumericTypes);
+
+TYPED_TEST(RandomNumericArrayTest, GenerateMinMax) {
+  auto field = this->GetField()->WithMetadata(
+      key_value_metadata({{"min", "0"}, {"max", "127"}, {"nan_probability", "0.0"}}));
+  auto batch = Generate({field}, 128, 0xDEADBEEF);
+  AssertSchemaEqual(schema({field}), batch->schema());
+  auto array = this->Downcast(batch->column(0));
+  auto it = array->begin();
+  while (it != array->end()) {
+    if ((*it).has_value()) {
+      ASSERT_GE(**it, typename TypeParam::c_type(0));
+      ASSERT_LE(**it, typename TypeParam::c_type(127));
+    }
+    it++;
+  }
+}
+
+TEST(TypeSpecificTests, FloatNan) {
+  auto field = arrow::field("float32", float32())
+                   ->WithMetadata(key_value_metadata({{"nan_probability", "1.0"}}));
+  auto batch = Generate({field}, 128, 0xDEADBEEF);
+  AssertSchemaEqual(schema({field}), batch->schema());
+  auto array = internal::checked_pointer_cast<NumericArray<FloatType>>(batch->column(0));
+  auto it = array->begin();
+  while (it != array->end()) {
+    if ((*it).has_value()) {
+      ASSERT_TRUE(std::isnan(**it));
+    }
+    it++;
+  }
+}
+
+TEST(TypeSpecificTests, RepeatedStrings) {
+  auto field =
+      arrow::field("string", utf8())->WithMetadata(key_value_metadata({{"unique", "1"}}));
+  auto batch = Generate({field}, 128, 0xDEADBEEF);
+  AssertSchemaEqual(schema({field}), batch->schema());
+  auto array = internal::checked_pointer_cast<StringArray>(batch->column(0));
+  auto it = array->begin();
+  util::optional<util::string_view> singular_value;
+  while (it != array->end()) {
+    if ((*it).has_value()) {
+      if (!singular_value.has_value()) {
+        singular_value = *it;
+      } else {
+        ASSERT_EQ(*singular_value, **it);
+      }
+    }
+    it++;
+  }
+}
+
```

Comment:
There are numerous untested options, please ensure each recognized metadata key is hit at least once